File: api/src/main/java/org/delivery/api/common/api/Result.java
package org.delivery.api.common.api;

import jakarta.persistence.criteria.CriteriaBuilder.In;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.delivery.api.common.error.ErrorCode;
import org.delivery.api.common.error.ErrorCodeInterface;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Result {

    private Integer resultCode;
    private String resultMessage;
    private String resultDescription;

    public static Result OK() {
        return Result.builder()
                .resultCode(ErrorCode.OK.getErrorCode())
                .resultMessage("OK")
                .resultDescription(ErrorCode.OK.getDescription())
                .build();
    }

    public static Result ERROR(ErrorCodeInterface errorCodeInterface) {
        return Result.builder()
                .resultCode(errorCodeInterface.getErrorCode())
                .resultMessage("ERROR")
                .resultDescription(errorCodeInterface.getDescription())
                .build();
    }

    public static Result ERROR(ErrorCodeInterface errorCodeInterface, Throwable throwable) {
        return Result.builder()
                .resultCode(errorCodeInterface.getErrorCode())
                .resultMessage(throwable.getLocalizedMessage())
                .resultDescription(errorCodeInterface.getDescription())
                .build();
    }

    public static Result ERROR(ErrorCodeInterface errorCodeInterface, String message) {
        return Result.builder()
                .resultCode(errorCodeInterface.getErrorCode())
                .resultMessage(message)
                .resultDescription(errorCodeInterface.getDescription())
                .build();
    }
}

File: common/src/main/kotlin/org/delivery/common/api/Result.kt
package org.delivery.common.api

import org.delivery.common.error.ErrorCode
import org.delivery.common.error.ErrorCodeInterface

data class Result (
    var resultCode: Int? = null,
    var resultMessage: String? = null,
    var resultDescription: String? = null,
){

    // 자바에서 static 메소드와 동일
    companion object {

        @JvmStatic
        fun OK() : Result {
            return Result(
                resultCode = ErrorCode.OK.getErrorCode(),
                resultMessage = "OK",
                resultDescription = ErrorCode.OK.getDescription()
            )           
        }

        @JvmStatic
        fun ERROR(errorCodeInterface: ErrorCodeInterface) : Result {
            return Result(
                resultCode = errorCodeInterface.getErrorCode(),
                resultMessage = "ERROR",
                resultDescription = errorCodeInterface.getDescription()
            )
        }

        @JvmStatic
        fun ERROR(errorCodeInterface: ErrorCodeInterface, throwable: Throwable) : Result {
            return Result(
                resultCode = errorCodeInterface.getErrorCode(),
                resultMessage = throwable.localizedMessage,
                resultDescription = errorCodeInterface.getDescription()
            )
        }

        @JvmStatic
        fun ERROR(errorCodeInterface: ErrorCodeInterface, message: String) : Result {
            return Result(
                resultCode = errorCodeInterface.getErrorCode(),
                resultMessage = message,
                resultDescription = errorCodeInterface.getDescription()
            )
        }
    }

}